flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcUtils.java

```diff
@@ -0,0 +1,2657 @@
+				keyRows[i] = new Object();
+				valueRows[i] = new Object();
+			}
+
+			// read map kes and values
+			readField(keyRows, -1, keyType, keys, null);
+			readField(valueRows, -1, valueType, values, null);
+
+			if (lengthVector == null) {
+				long[] lengthVectorMap = map.lengths;
+				int offset = 0;
+				if (fieldIdx == -1) {
+					for (int i = 0; i < rows.length; i++) {
+						HashMap<Object, Object> resultMap = new HashMap<Object, Object>();
+						for (int j = 0; j < lengthVectorMap[i]; j++) {
+							resultMap.put(keyRows[offset], valueRows[offset]);
+							offset++;
+						}
+						rows[i] = resultMap;
+					}
+				}
+				else {
+					for (int i = 0; i < rows.length; i++) {
+						HashMap<Object, Object> resultMap = new HashMap<Object, Object>();
+						for (int j = 0; j < lengthVectorMap[i]; j++) {
+							resultMap.put(keyRows[offset], valueRows[offset]);
+							offset++;
+						}
+						((Row) rows[i]).setField(fieldIdx, resultMap);
+					}
+				}
+			}
+			else { // list of map
+
+				Object[] nestedMap = new Object[rows.length];
+				for (int i = 0; i < rows.length; i++) {
+					nestedMap[i] = new Object();
+				}
+
+				long[] lengthVectorMap = map.lengths;
+				int offset = 0;
+				for (int i = 0; i < rows.length; i++) {
+					HashMap<Object, Object> resultMap = new HashMap<Object, Object>();
+					for (int j = 0; j < lengthVectorMap[i]; j++) {
+						resultMap.put(keyRows[offset], valueRows[offset]);
+						offset++;
+					}
+					nestedMap[i] = resultMap;
+				}
+
+				offset = 0;
+				int length;
+				Object[] temp;
+				if (fieldIdx == -1) {
+					for (int i = 0; i < rows.length; i++) {
+						length = (int) lengthVector[i];
+						if (length == 0) {
+							rows[i] = new Object[0];
+						} else {
+							temp = new Object[length];
+							System.arraycopy(nestedMap, offset, temp, 0, length);
+							offset = offset + length;
+							rows[i] = temp;
+						}
+					}
+				}
+				else {
+					for (int i = 0; i < rows.length; i++) {
+						length = (int) lengthVector[i];
+						if (length == 0) {
+							((Row) rows[i]).setField(fieldIdx, new Object[0]);
+						} else {
+							temp = new Object[length];
+							System.arraycopy(nestedMap, offset, temp, 0, length);
+							offset = offset + length;
+							((Row) rows[i]).setField(fieldIdx, temp);
+						}
+					}
+				}
+			}
+		}
+	}
+
+	private static void fillColumnWithRepeatingValue(Object[] rows, int fieldIdx, Object repeatingValue) {
+
+		if (fieldIdx == -1) {
+			for (int i = 0; i < rows.length; i++) {
+				rows[i] = repeatingValue;
+			}
+		}
+		else {
+			for (int i = 0; i < rows.length; i++) {
+				((Row) rows[i]).setField(fieldIdx, repeatingValue);
+			}
+		}
+	}
+
+	private static void fillListWithRepeatingValue(Object[] rows, int fieldIdx, Object repeatingValue, long[] lengthVector) {
+
+		int length;
+		if (fieldIdx == -1) {
+			for (int i = 0; i < rows.length; i++) {
+				length = (int) lengthVector[i];
+				Object[] temp = new Object[length];
```

Comment:
The array should have the correct type. We can try to do this with `Array.newInstance()` and a `getClassForField()` method, but I'm mot sure if it is possible to achieve this in a generic way for primitive types.